src/System.Memory/src/System/Buffers/SequenceReaderExtensions_binary.cs

```diff
@@ -0,0 +1,170 @@
+
+namespace System.Buffers
+{
+    public static partial class SequenceReaderExtensions
+    {
+        /// <summary>
+        /// Try to read the given type out of the buffer if possible.
+        /// </summary>
+        /// <remarks>
+        /// The read is unaligned.
+        /// </remarks>
+        /// <returns>
+        /// True if successful. <paramref name="value"/> will be default if failed.
+        /// </returns>
+        public static unsafe bool TryRead<T>(ref this SequenceReader<byte> reader, out T value) where T : unmanaged
+        {
+            ReadOnlySpan<byte> span = reader.UnreadSpan;
+            if (span.Length < sizeof(T))
+                return TryReadSlow(ref reader, out value);
+
+            value = Unsafe.ReadUnaligned<T>(ref MemoryMarshal.GetReference(span));
+            reader.Advance(sizeof(T));
+            return true;
+        }
+
+        private static unsafe bool TryReadSlow<T>(ref SequenceReader<byte> reader, out T value) where T : unmanaged
+        {
+            Debug.Assert(reader.UnreadSpan.Length < sizeof(T));
+
+            // Not enough data in the current segment, try to peek for the data we need.
+            byte* buffer = stackalloc byte[sizeof(T)];
+            Span<byte> tempSpan = new Span<byte>(buffer, sizeof(T));
+
+            if (reader.Peek(tempSpan).Length < sizeof(T))
+            {
+                value = default;
+                return false;
+            }
+
+            value = Unsafe.ReadUnaligned<T>(ref MemoryMarshal.GetReference(tempSpan));
+            reader.Advance(sizeof(T));
+            return true;
+        }
+
+        /// <summary>
+        /// Reads an <see cref="Int16"/> as little endian.
+        /// </summary>
+        /// <returns>False if there wasn't enough data for an <see cref="Int16"/>.</returns>
+        public static bool TryReadInt16LittleEndian(ref this SequenceReader<byte> reader, out short value)
+        {
+            if (BitConverter.IsLittleEndian)
+            {
+                return reader.TryRead(out value);
+            }
+
+            return TryReadReverseEndianness(ref reader, out value);
+        }
+
+        /// <summary>
+        /// Reads an <see cref="Int16"/> as big endian.
+        /// </summary>
+        /// <returns>False if there wasn't enough data for an <see cref="Int16"/>.</returns>
+        public static bool TryReadInt16BigEndian(ref this SequenceReader<byte> reader, out short value)
+        {
+            if (!BitConverter.IsLittleEndian)
+            {
+                return reader.TryRead(out value);
+            }
+
+            return TryReadReverseEndianness(ref reader, out value);
+        }
+
+        private static bool TryReadReverseEndianness(ref SequenceReader<byte> reader, out short value)
+        {
+            if (reader.TryRead(out value))
+            {
+                value = BinaryPrimitives.ReverseEndianness(value);
+                return true;
+            }
+
+            return false;
```

Comment:
Did you consider implementing this without the extra branch? e.g.
```C#
bool result = reader.TryRead(out value);
value = BinaryPrimitives.ReverseEndianness(value);
return result;
```
The trade-off would be based on whether you expect it to be common or not for TryRead to return false.

Reply:
Very rare to fail. I'll try it out.